File: test_calculator.cpp
#include "calculator.h"
#include <cassert>
#include <iostream>
#include <cmath>

// 简单的断言宏（用于测试）
#define ASSERT_EQUAL(a, b) \
    do { \
        if (std::abs((a) - (b)) > 1e-9) { \
            std::cerr << "测试失败: " << #a << " != " << #b << " (" << (a) << " != " << (b) << ")" << std::endl; \
            return false; \
        } \
    } while(0)

#define ASSERT_TRUE(condition) \
    do { \
        if (!(condition)) { \
            std::cerr << "测试失败: " << #condition << " 为假" << std::endl; \
            return false; \
        } \
    } while(0)

bool testAdd() {
    Calculator calc;
    ASSERT_EQUAL(calc.add(5, 3), 8);
    ASSERT_EQUAL(calc.add(-1, 1), 0);
    ASSERT_EQUAL(calc.add(0, 0), 0);
    std::cout << "✓ 加法测试通过" << std::endl;
    return true;
}

bool testSubtract() {
    Calculator calc;
    ASSERT_EQUAL(calc.subtract(10, 3), 7);
    ASSERT_EQUAL(calc.subtract(5, 5), 0);
    ASSERT_EQUAL(calc.subtract(0, 5), -5);
    std::cout << "✓ 减法测试通过" << std::endl;
    return true;
}

bool testMultiply() {
    Calculator calc;
    ASSERT_EQUAL(calc.multiply(5, 3), 15);
    ASSERT_EQUAL(calc.multiply(0, 100), 0);
    ASSERT_EQUAL(calc.multiply(-2, 3), -6);
    std::cout << "✓ 乘法测试通过" << std::endl;
    return true;
}

bool testDivide() {
    Calculator calc;
    ASSERT_EQUAL(calc.divide(10, 2), 5);
    ASSERT_EQUAL(calc.divide(9, 3), 3);
    ASSERT_EQUAL(calc.divide(1, 3), 0.333333333, 1e-6);
    std::cout << "✓ 除法测试通过" << std::endl;
    return true;
}

bool testDivideByZero() {
    Calculator calc;
    try {
        calc.divide(10, 0);
        std::cerr << "测试失败: 应该抛出除以零异常" << std::endl;
        return false;
    } catch (const std::runtime_error& e) {
        // 预期的异常
    }
    std::cout << "✓ 除以零测试通过" << std::endl;
    return true;
}

bool testHistory() {
    Calculator calc;
    calc.add(1, 2);
    calc.multiply(3, 4);
    ASSERT_EQUAL(calc.getHistorySize(), 2);
    std::cout << "✓ 历史记录测试通过" << std::endl;
    return true;
}

bool testClearHistory() {
    Calculator calc;
    calc.add(1, 1);
    calc.clearHistory();
    ASSERT_EQUAL(calc.getHistorySize(), 0);
    std::cout << "✓ 清空历史测试通过" << std::endl;
    return true;
}

int main() {
    std::cout << "=== 计算器单元测试 ===" << std::endl << std::endl;
    
    int passed = 0;
    int total = 7;
    
    if (testAdd()) passed++;
    if (testSubtract()) passed++;
    if (testMultiply()) passed++;
    if (testDivide()) passed++;
    if (testDivideByZero()) passed++;
    if (testHistory()) passed++;
    if (testClearHistory()) passed++;
    
    std::cout << std::endl << "测试结果: " << passed << "/" << total << " 通过" << std::endl;
    
    return (passed == total) ? 0 : 1;
}

File: test_calculator.py
#!/usr/bin/env python3
"""
计算器模块的单元测试
"""

import unittest
from calculator import Calculator


class TestCalculator(unittest.TestCase):
    """计算器测试类"""
    
    def setUp(self):
        """每个测试前的设置"""
        self.calc = Calculator()
    
    def test_add(self):
        """测试加法"""
        self.assertEqual(self.calc.add(5, 3), 8)
        self.assertEqual(self.calc.add(-1, 1), 0)
        self.assertEqual(self.calc.add(0, 0), 0)
    
    def test_subtract(self):
        """测试减法"""
        self.assertEqual(self.calc.subtract(10, 3), 7)
        self.assertEqual(self.calc.subtract(5, 5), 0)
        self.assertEqual(self.calc.subtract(0, 5), -5)
    
    def test_multiply(self):
        """测试乘法"""
        self.assertEqual(self.calc.multiply(5, 3), 15)
        self.assertEqual(self.calc.multiply(0, 100), 0)
        self.assertEqual(self.calc.multiply(-2, 3), -6)
    
    def test_divide(self):
        """测试除法"""
        self.assertEqual(self.calc.divide(10, 2), 5)
        self.assertEqual(self.calc.divide(9, 3), 3)
        self.assertAlmostEqual(self.calc.divide(1, 3), 0.333333, places=5)
    
    def test_divide_by_zero(self):
        """测试除以零的错误处理"""
        with self.assertRaises(ValueError):
            self.calc.divide(10, 0)
    
    def test_history(self):
        """测试计算历史"""
        self.calc.add(1, 2)
        self.calc.multiply(3, 4)
        history = self.calc.get_history()
        
        self.assertEqual(len(history), 2)
        self.assertIn("1 + 2 = 3", history)
        self.assertIn("3 × 4 = 12", history)
    
    def test_clear_history(self):
        """测试清空历史"""
        self.calc.add(1, 1)
        self.calc.clear_history()
        self.assertEqual(len(self.calc.get_history()), 0)


if __name__ == "__main__":
    unittest.main()